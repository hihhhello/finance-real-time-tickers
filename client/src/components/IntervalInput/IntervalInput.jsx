import _ from "lodash";
import React, { useEffect, useState } from "react";
import { useSelector } from "react-redux";
// import { ENTER } from "../../consts/keyNames";
import styles from "./interval-input.module.scss";
import { useActions } from "../../hooks/useActions";
import { toast } from "react-toastify";

export const IntervalInput = () => {
  const userSettings = useSelector((state) => state.userSettings);
  const { socket } = useSelector((state) => state.socket);
  const [intervalInput, setIntervalInput] = useState("");
  const { setUpdateInterval } = useActions();
  const intervalInputHandler = () => {
    if (intervalInput === userSettings.updateInterval) return;
    if (isNaN(+intervalInput)) {
      toast.warning("Interval input must be an integer", {
        toastId: "interval-input-warning",
      });
      setIntervalInput(userSettings.updateInterval);
      return;
    }
    if (intervalInput < 1000) {
      toast.warning("Interval input must be more than 0 ms", {
        toastId: "interval-input-warning",
      });
      setIntervalInput(userSettings.updateInterval);
      return;
    }
    setUpdateInterval({ value: intervalInput, socket });
  };
  useEffect(() => {
    if (userSettings.updateInterval) {
      setIntervalInput(userSettings.updateInterval);
    }
  }, [userSettings.updateInterval]);
  return (
    <div className={styles.wrapper}>
      {_.isEmpty(userSettings) && !socket ? (
        <span className={styles.loading}>Loading...</span>
      ) : (
        <>
          <div className={styles.input}>
            <label htmlFor="update-interval">Update interval (ms)</label>
            <input
              type="text"
              id="update-interval"
              value={intervalInput}
              onChange={(e) => setIntervalInput(e.target.value)}
              // onKeyDown={(e) => intervalInputHandler(e)}
            />
          </div>
          <button className={styles.btn} onClick={intervalInputHandler}>
            set
          </button>
        </>
      )}
    </div>
  );
};
