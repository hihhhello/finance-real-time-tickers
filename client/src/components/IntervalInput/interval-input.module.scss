@import "../../scss/var";
.wrapper {
  font-size: 2.4rem;
  padding-top: 0.9em;
  width: 100%;
  max-width: 15em;
}

.input {
  display: flex;
  flex-direction: column;
  input {
    font-size: 1em;
    max-width: 8em;
  }
}

.btn {
  margin-top: 0.5em;
  font-size: 0.8em;
  padding: 0.2em 0.8em;
  text-transform: uppercase;
  border-radius: 0.6em;
  position: relative;
  top: 0;
  background-color: $success-color;
  color: #fff;
  border: none;
  cursor: pointer;
  transition: 0.3s all;
}

.loading {
  compose: placeholder from "../AvalTickers/aval-tickers.module.scss";
}
