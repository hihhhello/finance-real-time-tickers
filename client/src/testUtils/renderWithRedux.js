import { Provider } from "react-redux";
import { combineReducers, createStore } from "redux";
import { render } from "@testing-library/react";
import reducers from "../store/slices";

export const renderWithRedux = (
  component,
  {
    initialState,
    store = createStore(combineReducers(reducers), initialState),
  } = {}
) => {
  return {
    ...render(<Provider store={store}>{component}</Provider>),
    store,
  };
};
