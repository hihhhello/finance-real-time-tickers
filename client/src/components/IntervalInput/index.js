import { IntervalInput } from "./IntervalInput";
export default IntervalInput;
