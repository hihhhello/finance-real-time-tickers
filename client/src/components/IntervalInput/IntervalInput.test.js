import { screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { renderWithRedux } from "../../testUtils/renderWithRedux";
import { IntervalInput } from "./IntervalInput";

const initialState = {
  socket: null,
  userSettings: {
    updateInterval: 5000,
  },
};

describe("Interval input element", () => {
  it("renders with loader", () => {
    const { asFragment, getByText } = renderWithRedux(<IntervalInput />);

    expect(getByText("Loading...")).toBeInTheDocument();
    expect(asFragment(<IntervalInput />)).toMatchSnapshot();
  });

  it("renders", () => {
    const { asFragment, getByRole } = renderWithRedux(<IntervalInput />, {
      initialState: {
        ...initialState,
        socket: {},
      },
    });
    expect(getByRole("button", { name: /set/i })).toBeInTheDocument();
    expect(getByRole("textbox")).toHaveAttribute(
      "value",
      initialState.updateInterval
    );
    expect(asFragment(<IntervalInput />)).toMatchSnapshot();
  });

  it("insert new update interval", () => {
    const { getByRole } = renderWithRedux(<IntervalInput />, {
      initialState: {
        ...initialState,
        socket: {},
      },
    });
    expect(getByRole("textbox")).toHaveAttribute(
      "value",
      initialState.updateInterval
    );
    userEvent.type(
      getByRole("textbox"),
      "{backspace}{backspace}{backspace}{backspace}8000"
    );
    expect(getByRole("textbox")).toHaveAttribute("value", "8000");
  });

  // it("on click button updating interval value in user settings", async () => {
  //   const { getByRole } = renderWithRedux(<IntervalInput />, {
  //     initialState: {
  //       ...initialState,
  //     },
  //   });
  //   userEvent.click(getByRole("button", { name: /set/ }));
  // });
});
