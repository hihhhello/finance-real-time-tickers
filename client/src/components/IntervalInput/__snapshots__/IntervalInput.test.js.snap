// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Interval input element renders 1`] = `
<DocumentFragment>
  <div
    class="wrapper"
  >
    <div
      class="input"
    >
      <label
        for="update-interval"
      >
        Update interval (ms)
      </label>
      <input
        id="update-interval"
        type="text"
        value="5000"
      />
    </div>
    <button
      class="btn"
    >
      set
    </button>
  </div>
</DocumentFragment>
`;

exports[`Interval input element renders with loader 1`] = `
<DocumentFragment>
  <div
    class="wrapper"
  >
    <span
      class="loading"
    >
      Loading...
    </span>
  </div>
</DocumentFragment>
`;
